radix-engine/src/state_manager/staging.rs

```diff
@@ -0,0 +1,216 @@
+    fn new(parent_id: u64) -> Self {
+        StagedExecutionStoreNode {
+            parent_id,
+            locked: false,
+            spaces: BTreeMap::new(),
+            outputs: BTreeMap::new(),
+        }
+    }
+}
+
+/// Structure which manages the acyclic graph
+pub struct StagedExecutionStores<'s, S: ReadableSubstateStore + WriteableSubstateStore> {
+    parent: &'s mut S,
+    nodes: HashMap<u64, StagedExecutionStoreNode>,
+    cur_id: u64,
+}
+
+impl<'s, S: ReadableSubstateStore + WriteableSubstateStore> StagedExecutionStores<'s, S> {
+    pub fn new(parent: &'s mut S) -> Self {
+        StagedExecutionStores {
+            parent,
+            nodes: HashMap::new(),
+            cur_id: 0,
+        }
+    }
+
+    pub fn new_child_node(&mut self, parent_id: u64) -> u64 {
+        if parent_id != 0 {
+            let parent = self.nodes.get_mut(&parent_id).unwrap();
+            parent.locked = true;
+        }
+
+        self.cur_id = self.cur_id + 1;
+        self.nodes
+            .insert(self.cur_id, StagedExecutionStoreNode::new(parent_id));
+        self.cur_id
+    }
+
+    pub fn get_output_store<'t>(&'t mut self, id: u64) -> ExecutionStore<'t, 's, S> {
+        if id != 0 && self.nodes.get(&id).unwrap().locked {
+            panic!("Should not write to locked node");
+        }
+
+        ExecutionStore { stores: self, id }
+    }
+
+    fn remove_children(&mut self, id: u64) {
+        let mut to_delete = Vec::new();
+        for (to_delete_id, node) in &self.nodes {
+            if node.parent_id == id {
+                to_delete.push(*to_delete_id);
+            }
+        }
+        for to_delete_id in to_delete {
+            self.remove_children(to_delete_id);
+            self.nodes.remove(&to_delete_id);
+        }
+    }
+
+    fn set_root_parent(&mut self, id: u64) {
+        for node in self.nodes.values_mut().filter(|node| id == node.parent_id) {
+            node.parent_id = 0;
+        }
+    }
+
+    pub fn merge_to_parent(&mut self, id: u64) {
+        self.merge_to_parent_recurse(id, false)
+    }
+
+    fn merge_to_parent_recurse(&mut self, id: u64, remove_children: bool) {
+        if id == 0 {
+            if remove_children {
+                self.remove_children(0);
+            }
+            return;
+        }
+
+        let node = self.nodes.remove(&id).unwrap();
+        if remove_children {
+            self.remove_children(id);
+        }
+
+        self.merge_to_parent_recurse(node.parent_id, true);
+
+        for (address, output_id) in node.spaces {
+            self.parent.put_space(&address, output_id);
+        }
+
+        for (address, output) in node.outputs {
+            self.parent.put_substate(&address, output);
+        }
+
+        if !remove_children {
+            self.set_root_parent(id);
+        }
+    }
+}
+
+pub struct ExecutionStore<'t, 's, S: ReadableSubstateStore + WriteableSubstateStore> {
+    stores: &'t mut StagedExecutionStores<'s, S>,
+    id: u64,
+}
```

Comment:
Looks like `ExecutionStore` can be the node.

Reply:
Can't because we need a reference to the entire store.

Reply:
Sounds like resolvable by a bi-directional tree structure, but we're in Rust not Java.  It's going to be ugly to fight with the ownership system, http://way-cooler.org/blog/2016/08/14/designing-a-bi-mutable-directional-tree-safely-in-rust.html

Kindly ignore my comment.

Reply:
Yup, definitely different with Rust.